lib/segment/src/index/field_index/geo_index/immutable_geo_index.rs

```diff
@@ -0,0 +1,237 @@
+    }
+
+    pub fn get_points_per_hash(&self) -> impl Iterator<Item = (&GeoHash, usize)> {
+        self.counts_per_hash
+            .iter()
+            .map(|counts| (&counts.hash, counts.points as usize))
+    }
+
+    pub fn get_points_of_hash(&self, hash: &GeoHash) -> usize {
+        if let Ok(index) = self.counts_per_hash.binary_search_by(|x| x.hash.cmp(hash)) {
+            self.counts_per_hash[index].points as usize
+        } else {
+            0
+        }
+    }
+
+    pub fn get_values_of_hash(&self, hash: &GeoHash) -> usize {
+        if let Ok(index) = self.counts_per_hash.binary_search_by(|x| x.hash.cmp(hash)) {
+            self.counts_per_hash[index].values as usize
+        } else {
+            0
+        }
+    }
+
+    pub fn load(&mut self) -> OperationResult<bool> {
+        let mut mutable_geo_index = MutableGeoMapIndex::new(
+            self.db_wrapper.database.clone(),
+            &self.db_wrapper.column_name,
+        );
+        let result = mutable_geo_index.load()?;
+
+        let MutableGeoMapIndex {
+            points_per_hash,
+            values_per_hash,
+            points_map,
+            point_to_values,
+            points_count,
+            points_values_count,
+            max_values_per_point,
+            ..
+        } = mutable_geo_index;
+
+        let mut counts_per_hash: BTreeMap<GeoHash, Counts> = Default::default();
+        for (hash, points) in points_per_hash {
+            counts_per_hash.insert(
+                hash.clone(),
+                Counts {
+                    hash,
+                    points: points as u32,
+                    values: 0,
+                },
+            );
+        }
+        for (hash, values) in values_per_hash {
+            if let Some(counts) = counts_per_hash.get_mut(&hash) {
+                counts.values = values as u32;
+            } else {
+                counts_per_hash.insert(
+                    hash.clone(),
+                    Counts {
+                        hash,
+                        points: 0,
+                        values: values as u32,
+                    },
+                );
+            }
+        }
+
+        self.counts_per_hash = counts_per_hash.values().cloned().collect();
+        self.points_map = points_map
+            .iter()
+            .map(|(k, v)| (k.clone(), v.clone()))
+            .collect();
+        self.point_to_values = ImmutablePointToValues::new(point_to_values);
+        self.points_count = points_count;
+        self.points_values_count = points_values_count;
+        self.max_values_per_point = max_values_per_point;
+
+        Ok(result)
+    }
+
+    pub fn remove_point(&mut self, idx: PointOffsetType) -> OperationResult<()> {
+        let removed_geo_points = self.point_to_values.remove_point(idx);
+        if removed_geo_points.is_empty() {
+            return Ok(());
+        }
+
+        self.points_count -= 1;
+        self.points_values_count -= removed_geo_points.len();
```

Comment:
We probably want to subtract counts here before returning early.

Reply:
In geo index, points count is a count of points with non-empty values. Not sure that it's the correct behavior but I kept this logic as is for backward compatibility